XenAdmin/Wizards/PatchingWizard/PatchingWizard_AutomatedUpdatesPage.cs

```diff
@@ -290,68 +276,133 @@ private void UpdateStatusTextBox()
         private void WorkerDoWork(object sender, DoWorkEventArgs doWorkEventArgs)
         {
             var bgw = sender as UpdateProgressBackgroundWorker;
+            PlanAction action = null;
 
-            foreach (PlanAction action in bgw.AllActions)
+            try
             {
-                try
+                //running actions (non-delayed)
+                foreach (var a in bgw.PlanActions)
                 {
+                    action = a;
+
                     if (bgw.CancellationPending)
                     {
                         doWorkEventArgs.Cancel = true;
                         return;
                     }
 
-                    InitializePlanAction(bgw, action);
+                    RunPlanAction(bgw, action);
+                }
+
+                // running delayed actions, but skipping the ones that should be skipped
+                // iterating through hosts
+                foreach (var kvp in bgw.DelayedActionsByHost)
+                {
+                    var h = kvp.Key;
+                    var actions = kvp.Value;
+
+                    //run all restart-alike plan actions
+                    foreach (var a in actions.Where(a => a.IsRestartRelatedPlanAction()))
+                    {
+                        action = a;
+
+                        if (bgw.CancellationPending)
+                        {
+                            doWorkEventArgs.Cancel = true;
+                            return;
+                        }
+
+                        RunPlanAction(bgw, action);
+                    }
+
+                    //run the rest
+                    foreach (var a in actions.Where(a => !a.IsRestartRelatedPlanAction()))
+                    {
+                        action = a;
 
-                    bgw.ReportProgress(0, action);
-                    action.Run();
+                        if (bgw.CancellationPending)
+                        {
+                            doWorkEventArgs.Cancel = true;
+                            return;
+                        }
 
-                    Thread.Sleep(1000);
+                        // any non-restart-alike delayed action needs to be run if:
+                        // - this host is pre-Ely and there isn't any restart plan action among the delayed actions, or
+                        // - this host is Ely or above and bgw.AvoidRestartHosts contains the host's uuid (shows that live patching must have succeeded)
+                        if (!Helpers.ElyOrGreater(h) && !actions.Any(pa => pa.IsRestartRelatedPlanAction())
+                            || Helpers.ElyOrGreater(h) && bgw.AvoidRestartHosts != null && bgw.AvoidRestartHosts.Contains(h.uuid))

```

Comment:
Here we also need to include the case where there wasn't any restart related action in the list (!actions.Any(pa => pa.IsRestartRelatedPlanAction())